### ISSUE TITLE
[question] different behavior when using "tool_requires" in Conan 1.X and Conan 2.X (only in test-package)

### ISSUE THREAD

**Issue by DominikMoreira** [conan-io/conan#15165]
### What is your question?

**Description:**
When utilizing tool_requires in a test-package with Conan 1.x, I observe a different behavior compared to Conan 2.x.

**Used profiles:** 
Build profile: 
```
[settings]
arch=x86_64
os=Linux
```

Host-profile:
```
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
os=Linux
os_build=Linux
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=10.3
```

**Conan 1.x Behavior:**
Steps to Reproduce:

- conan create . randomUser/randomChannel -pr:b build-profile -pr:h host-profile
- Using build and host profiles, as outlined above, results in the following output:

`ERROR: protoc/3.18.0_17@randomUser/randomChannel: 'settings.compiler' value not defined`

**Conan 2.x Behavior:**
Steps to Reproduce: 

- conan2 create . --user randomUser --channel randomChannel -pr:b build-profile -pr:h host-profile
- Build successful

With the same profiles, the tool_requires functionality works as expected.

**Conclusion**:
In Conan 1.x it seems to build protoc from the test package. The tool_requires will lead to switching the build and host profiles and therefore not finding a compiler.setting (because we are using a minimal build profile).

In Conan 2.x on the other hand it first build only the protoc package and then the test consumes the created package (as expected). 

To summarize the observations the following question occured: 
* Is this behavior expected?
* Is there a better solution than using "requires" instead "tool_requires"?

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @DominikWunderlich 

Thanks for your question.

The behavior is indeed different in Conan 1.X and Conan 2.0. The main difference is that Conan 1.X computes the graph starting from ``test_package``, but Conan 2.0 computes the graph starting from the package ``conanfile.py`` and only once the package has been built, it will launch the ``test_package`` step.

Some important remarks:
- When ``conan create`` and ``test_package`` of something that is intended to run as a ``tool_require`` the correct approach is to define the ``conan create ... --build-require`` argument. That argument correctly creates and put the dependency in the "host" context, to accept the correct settings
- If you do that, you'll most likely get the same error, because the compiler is not defined in your build context
- In your current scenario was kind of ok to build for the "host" context, because it is not really a cross-build scenario, but it is conceptually incorrect, as it would fail in cross-build, and it will have incorrect results if you have different configurations in build and host context, for example "shared" or static library with ``option`` ``shared=True``, or using ``Release`` or ``Debug`` build type.

**Comment by memsharded** [comment]
Closing the ticket as responded, please create new tickets for any further question or issue, thanks!
